src/execution/operator/join/physical_hash_join.cpp

```diff
@@ -1192,15 +1190,15 @@ SourceResultType PhysicalHashJoin::GetData(ExecutionContext &context, DataChunk
 		if (!lstate.TaskFinished() || gstate.AssignTask(sink, lstate)) {
 			lstate.ExecuteTask(sink, gstate, chunk);
 		} else {
-			lock_guard<mutex> guard(gstate.lock);
+			auto guard = gstate.Lock();
 			if (gstate.TryPrepareNextStage(sink) || gstate.global_stage == HashJoinSourceStage::DONE) {
-				for (auto &state : gstate.blocked_tasks) {
-					state.Callback();
-				}
-				gstate.blocked_tasks.clear();
+				gstate.UnblockTasks(guard);
 			} else {
-				gstate.blocked_tasks.push_back(input.interrupt_state);
-				return SourceResultType::BLOCKED;
+				if (gstate.BlockTask(guard, input.interrupt_state)) {
```

Comment:
Perhaps we can add a `BlockSource` and `BlockSink` that directly return this, so we can do `return gstate.BlockSource(...)` here?